tx_service.go

```diff
@@ -3,53 +3,74 @@ package bux
 import (
 	"context"
 	"encoding/hex"
-	"errors"
+	"fmt"
 
-	"github.com/BuxOrg/bux/chainstate"
 	"github.com/BuxOrg/bux/utils"
-	"github.com/mrz1836/go-datastore"
 )
 
-// processTransactions will process transaction records
-func processTransactions(ctx context.Context, maxTransactions int, opts ...ModelOps) error {
-	queryParams := &datastore.QueryParams{
-		Page:          1,
-		PageSize:      maxTransactions,
-		OrderByField:  "created_at",
-		SortDirection: "asc",
-	}
+// The transaction is treat as external incoming transaction - transaction without a draft
+// Only use this function when you know what you are doing!
+func saveRawTransaction(ctx context.Context, c ClientInterface, allowUnknown bool, txHex string, opts ...ModelOps) (*Transaction, error) {
+	newOpts := c.DefaultModelOptions(append(opts, New())...)
+	tx := newTransaction(txHex, newOpts...)
 
-	conditions := map[string]interface{}{
-		"$or": []map[string]interface{}{{
-			blockHeightField: 0,
-		}, {
-			blockHeightField: nil,
-		}},
+	// Ensure that we have a transaction id (created from the txHex)
+	id := tx.GetID()
+	if len(id) == 0 {
+		return nil, ErrMissingTxHex
 	}
 
-	records := make([]Transaction, 0)
-	err := getModelsByConditions(ctx, ModelTransaction, &records, nil, &conditions, queryParams, opts...)
+	// Create the lock and set the release for after the function completes
+	unlock, err := newWriteLock(
+		ctx, fmt.Sprintf(lockKeyRecordTx, id), c.Cachestore(),
+	)
+	defer unlock()
 	if err != nil {
-		return err
-	} else if len(records) == 0 {
-		return nil
+		return nil, err
 	}
 
-	txs := make([]*Transaction, 0)
-	for index := range records {
-		records[index].enrich(ModelTransaction, opts...)
-		txs = append(txs, &records[index])
+	// Logic moved from BeforeCreating hook - should be refactorized in next iteration
+
+	if !c.IsITCEnabled() && !tx.hasOneKnownDestination(ctx, c) {
```

Comment:
Why we check known destinations only if ITC (incoming transaction check) is disabled? Shouldn't that be checked only if it is enabled or something?

Reply:
It was leftover from `BeforeCreate` hook and make no sense here - thank for noticing this